### ISSUE TITLE
TCK dynamic tests need unique test names

### ISSUE THREAD

**Issue by andymc12** [eclipse/microprofile-graphql#97]
When running the TCK, failures in the dynamic tests show up as the same test method name in the summary - something like:
```
Test Failures: 
  testSpecification
  testSpecification
  testSpecification
  testSpecification
```

It would be nice to see more specific test name - maybe something like `testSpecification_jsonbPropsTest` ?

**Comment by phillip-kruger** [comment]
Hi @andymc12 - what version of surefire are you using ? (I assume you run the TCK test from maven ?)

**Comment by phillip-kruger** [comment]
Hi @andymc12 - So my TCK always gave me more information than that you reported. I am trying to clean it up a bit, as it's hard to follow the test case that has an error. 

For the Dynamic Schema test it will now print like this:

```
[INFO]   Run 1: PASS
[ERROR]   Run 2: SchemaDynamicValidityTest>Arquillian.run:138->testPartsOfSchema:87 [testIgnoreOnFieldExcludedFromInputAndOutputTypes] (2) - Found field "invisible" that should be ignored" expected [false] but found [true]
[INFO]   Run 3: PASS
```
... etc

So when it fails, you can get the testName as defined in the csv and the number:

Example above:

Test Name: testIgnoreOnFieldExcludedFromInputAndOutputTypes
in the csv:
```
# testIgnoreOnFieldExcludedFromInputAndOutputTypes
1| type Item           |   powerLevel                                               |   Missing expected, un-ignored field, "powerLevel"
2| type Item           |   !invisible                                               |   Found field "invisible" that should be ignored"
3| input ItemInput     |   powerLevel                                               |   Missing expected, un-ignored field, "powerLevel"
4| input ItemInput     |   !invisible                                               |   Found field "invisible" that should be ignored"
```

Line 2 is the test failing.

I am thinking that this is just in my smallrye impl that it prints like this ? Can you let me know ? Thanks

**Comment by phillip-kruger** [comment]
Hi @andymc12  - I enhanced the test further to just print the name and expected in and received out when running the test, something like this:
```
[ERROR]   Run 3: GraphQLDynamicClientTest>Arquillian.run:138->testSpecification:79->runTest:113 setRivalTeam data
Expected: a JSON object
     got: null
```

And if the test failed there will be a log file in the target folder with all the details, example addHeroToTeam.log :
```

============= addHeroToTeam =============

errorMessage = addHeroToTeam data
Expected: a JSON object
     got: null


given input = mutation addHeroToTeam {
    addHeroToTeam(hero: "Starlord", team: "Avengers") {
        name
        members {
            name
        }
    }
}

variables input = null

http headers input = null

expected output = 
{
    "data": {
        "addHeroToTeam": {
            "name": "Avengers",
            "members": [
                {
                    "name": "Iron Man"
                },
                {
                    "name": "Spider Man"
                },
                {
                    "name": "Starlord"
                },
                {
                    "name": "Wolverine"
                }
            ]
        }
    }
}

received output = 
{
    "errors": [
        {
            "message": "Validation error of type UnknownArgument: Unknown field argument hero @ 'addHeroToTeam'",
            "locations": [
                {
                    "line": 2,
                    "column": 19
                }
            ],
            "extensions": {
                "classification": "ValidationError"
            }
        },
        {
            "message": "Validation error of type UnknownArgument: Unknown field argument team @ 'addHeroToTeam'",
            "locations": [
                {
                    "line": 2,
                    "column": 37
                }
            ],
            "extensions": {
                "classification": "ValidationError"
            }
        }
    ],
    "data": null
}
```
I am doing a PR for this. Let me know.

**Comment by andymc12** [comment]
Hi @phillip-kruger - thanks for working on this - I never suspected it could be the surefire plugin version.  We are using 2.17 - that sounds like it may be a bit old.  I see a comment in there saying something about issues on z/OS with newer versions, so that may be why we're using this particular version.  I'll do some testing with newer versions to see if I get the same results you are.

**Comment by andymc12** [comment]
I tried changing the version of surefire to 2.22.2 (the latest GA version) and 3.0.0-M3 (the latest pre-release), and didn't see any difference in the reporting. 

Looking into it a little more, I think the problem is in the way that the OpenLiberty build tooling reports the failures.  When I look at the Maven output directly, I see results like this:

```
[INFO]   Run 16: PASS
[ERROR]   Run 17: GraphQLDynamicClientTest>Arquillian.run:164->testSpecification:89->runTest:122 ============= ex_partialResultsInGraphQLExceptionSubclass =============

given input = query allHeroesWithSpecificError {
...
```

But OpenLiberty uses Gradle for the build and has a Maven "runner" class.  The output from Gradle looks like this:
```
    [junit] Running org.eclipse.microprofile.graphql.tck.FATSuite
    [junit] 
    [junit] TCK TESTS THAT DID NOT PASS:
    [junit]                                testSpecification
    [junit]                                testSpecification
    [junit] 
    [junit] 
    [junit] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 27.466 sec
     [echo] Finished running org.eclipse.microprofile.graphql.tck.FATSuite
     [echo] XML report is available at /Users/andymc/dev/libertyGit/open-liberty/dev/com.ibm.ws.microprofile.graphql_fat_tck/build/libs/autoFVT/results/junit
```

The other thing to note is that OpenLiberty uses JUnit for test result reporting, so it converts the TestNG results to JUnit results file - and that produces XML like this:

```
  <testcase classname="org.eclipse.microprofile.graphql.tck.dynamic.GraphQLDynamicClientTest" name="testSpecification" time="0.020"/>
  <testcase classname="org.eclipse.microprofile.graphql.tck.dynamic.GraphQLDynamicClientTest" name="testSpecification" time="0.030"/>
  <testcase classname="org.eclipse.microprofile.graphql.tck.dynamic.GraphQLDynamicClientTest" name="testSpecification" time="0.022"/>
  <testcase classname="org.eclipse.microprofile.graphql.tck.dynamic.GraphQLDynamicClientTest" name="testSpecification" time="0.019">
    <failure message="============= ex_partialResultsInGraphQLException =============  given input = query allHeroesWithError 
...
```

So the details in the failure message indicate which test failed, but when these results gets reported on the build page, only the test case classname and test case name are listed.  If users download the full test results, then they will see the failure message (and they can also see the Maven output, TestNG results, etc.).  

Since this seems to be a problem unique to OpenLiberty's build process/reporting, I don't think this should be prioritized very high.  If there is an easy way to replace the test case name (i.e. "testSpecification") with the actual dynamic test name (i.e. "ex_partialResultsInGraphQLException"), then I'd say let's do that.  If not, then we can probably close this issue and I can open an issue in the OpenLiberty project to do some sort of post-processing to get better results-publishing.

Thanks for your help with this, @phillip-kruger!

**Comment by phillip-kruger** [comment]
This is done as part of PR #99
